src/hdmf/common/table.py

```diff
@@ -0,0 +1,611 @@
+class DynamicTable(Container):
+    r"""
+    A column-based table. Columns are defined by the argument *columns*. This argument
+    must be a list/tuple of :class:`~hdmf.common.table.VectorData` and :class:`~hdmf.common.table.VectorIndex` objects
+    or a list/tuple of dicts containing the keys ``name`` and ``description`` that provide the name and description
+    of each column in the table. Additionally, the keys ``index`` and ``table`` for specifying additional structure to
+    the table columns. Setting the key ``index`` to ``True`` can be used to indicate that the
+    :class:`~hdmf.common.table.VectorData` column will store a ragged array (i.e. will be accompanied with a
+    :class:`~hdmf.common.table.VectorIndex`). Setting the key ``table`` to ``True`` can be used to indicate that the
+    column will store regions to another DynamicTable.
+
+    Columns in DynamicTable subclasses can be statically defined by specifying the class attribute *\_\_columns\_\_*,
+    rather than specifying them at runtime at the instance level. This is useful for defining a table structure
+    that will get reused. The requirements for *\_\_columns\_\_* are the same as the requirements described above
+    for specifying table columns with the *columns* argument to the DynamicTable constructor.
+    """
+
+    __fields__ = (
+        {'name': 'id', 'child': True},
+        {'name': 'columns', 'child': True},
+        'colnames',
+        'description'
+    )
+
+    __columns__ = tuple()
+
+    @ExtenderMeta.pre_init
+    def __gather_columns(cls, name, bases, classdict):
+        '''
+        This classmethod will be called during class declaration in the metaclass to automatically
+        include all columns declared in subclasses
+        '''
+        if not isinstance(cls.__columns__, tuple):
+            msg = "'__columns__' must be of type tuple, found %s" % type(cls.__columns__)
+            raise TypeError(msg)
+
+        if len(bases) and 'DynamicTable' in globals() and issubclass(bases[-1], Container) \
+                and bases[-1].__columns__ is not cls.__columns__:
+            new_columns = list(cls.__columns__)
+            new_columns[0:0] = bases[-1].__columns__
+            cls.__columns__ = tuple(new_columns)
+
+    @docval({'name': 'name', 'type': str, 'doc': 'the name of this table'},    # noqa: C901
+            {'name': 'description', 'type': str, 'doc': 'a description of what is in this table'},
+            {'name': 'id', 'type': ('array_data', ElementIdentifiers), 'doc': 'the identifiers for this table',
+             'default': None},
+            {'name': 'columns', 'type': (tuple, list), 'doc': 'the columns in this table', 'default': None},
+            {'name': 'colnames', 'type': 'array_data', 'doc': 'the names of the columns in this table',
+             'default': None})
+    def __init__(self, **kwargs):
+        id, columns, desc, colnames = popargs('id', 'columns', 'description', 'colnames', kwargs)
```

Comment:
The init for DynamicTable is quite complex. It would be nice to add some documentation to indicate what happens where.